src/Builder/AbstractFormContractor.php

```diff
@@ -0,0 +1,197 @@
+use Sonata\AdminBundle\Form\Type\ModelReferenceType;
+use Sonata\AdminBundle\Form\Type\ModelType;
+use Sonata\AdminBundle\Form\Type\ModelTypeList;
+use Sonata\Form\Type\CollectionType;
+use Symfony\Component\Form\Extension\Core\Type\FormType;
+use Symfony\Component\Form\FormFactoryInterface;
+
+abstract class AbstractFormContractor implements FormContractorInterface
+{
+    /**
+     * @var FormFactoryInterface
+     */
+    protected $formFactory;
+
+    public function __construct(FormFactoryInterface $formFactory)
+    {
+        $this->formFactory = $formFactory;
+    }
+
+    public function fixFieldDescription(AdminInterface $admin, FieldDescriptionInterface $fieldDescription)
+    {
+        $fieldDescription->setOption('edit', $fieldDescription->getOption('edit', 'standard'));
+
+        // NEXT_MAJOR: Change "$this->hasAssociation($fieldDescription)" with: "$fieldDescription->describesAssociation()".
+        if ($this->hasAssociation($fieldDescription) || $fieldDescription->getOption('admin_code')) {
+            $admin->attachAdminClass($fieldDescription);
+        }
+    }
+
+    /**
+     * @return FormFactoryInterface
+     */
+    public function getFormFactory()
+    {
+        return $this->formFactory;
+    }
+
+    public function getFormBuilder($name, array $formOptions = [])
+    {
+        return $this->getFormFactory()->createNamedBuilder($name, FormType::class, null, $formOptions);
+    }
+
+    public function getDefaultOptions($type, FieldDescriptionInterface $fieldDescription)
+    {
+        // NEXT_MAJOR: Remove this line and update the function signature.
+        $formOptions = \func_get_args()[2] ?? [];
+
+        $options = [];
+        $options['sonata_field_description'] = $fieldDescription;
+
+        if ($this->isAnyInstanceOf($type, [
+            ModelType::class,
+            ModelTypeList::class,
+            ModelListType::class,
+            ModelHiddenType::class,
+            ModelAutocompleteType::class,
+            ModelReferenceType::class,
+        ])) {
+            // NEXT_MAJOR: Remove this check.
+            if ('list' === $fieldDescription->getOption('edit')) {
+                throw new \LogicException(sprintf(
+                    'The `%s` type does not accept an `edit` option anymore,'
+                    .' please review the UPGRADE-2.1.md file from the SonataAdminBundle',
+                    ModelType::class
+                ));
+            }
+
+            $options['class'] = $fieldDescription->getTargetModel();
+            $options['model_manager'] = $fieldDescription->getAdmin()->getModelManager();
+
+            if ($this->isAnyInstanceOf($type, [ModelAutocompleteType::class])) {
+                if (!$fieldDescription->getAssociationAdmin()) {
+                    throw new \RuntimeException(sprintf(
+                        'The current field `%s` is not linked to an admin.'
+                        .' Please create one for the target model: `%s`',
```

Comment:
```suggestion
                        .' Please create one for the target model: `%s`.',
```

Reply:
Do you agree with this suggestion @franmomu?

Reply:
sure! sorry I thought I had changed it.